### ISSUE TITLE
Introduce a generic `Usage` type to enforce ordered deletion

### ISSUE THREAD

**Issue by negz** [crossplane/crossplane#3393]
### What problem are you facing?

Crossplane is built on Kubernetes, which leans into eventual consistency. When several things need to happen in a particular order the “Kubernetes way” is to specify the desired end state and rely on the relevant controllers to ensure the system eventually arrives at that state, largely through the use of constant reconciliation (aka retries).

Take for example the creation of a VPC Network and a Subnet. The VPC must be created and ready before a Subnet can be created within it. In Kubernetes it’s typical to request both resources be created at once (thereby declaring the desired state). The creation of the Subnet will simply fail until its dependency - the VPC - becomes available.

This loosely coupled, eventually consistent approach is simple and resilient but it can appear chaotic. Operations must be attempted more times than would be necessary in a “smarter” system, and often the temporarily failing operations will emit warnings that may be confusing or concerning to the uninitiated. Some other systems - notably Terraform - avoid this by computing a dependency graph before attempting to enact the desired state. This allows the system to avoid trying operations it knows will fail due to missing dependencies until those dependencies are expected to be satisfied.

A frequent negative side-effect of the eventually consistent approach that we see in Crossplane is resources becoming orphaned at delete time because their successful deletion depends on the existence of other resources which are often deleted before their dependents. It’s possible to author a set of interdependent resources that will eventually create successfully, but may not eventually delete successfully.

Two variants of this problem are:

* **Orphaned Managed Resources (MRs)**. Consider a Helm Release MR that is deployed to an EKS Cluster MR. The Helm Release has a dependency on the EKS Cluster’s connection details - it uses them as Provider credentials. If the EKS Cluster is deleted before the Helm Release the Helm Release controller becomes unable to connect to its Provider in order to delete the Release. In practice the Release External Resource (ER) was implicitly deleted along with the GKE Cluster - there’s actually no work left for the Helm release controller to do - but it cannot know that. This results in the Helm Release MR being ‘orphaned’ in a pending deletion that can never be satisfied.
* **Orphaned ‘Side-Effect’ Resources**. Consider the same example as above, and assume the Helm Release includes a Kubernetes Service resource of type: LoadBalancer. A side-effect of the creation of this Service will be to create an ELB in the EKS cluster’s VPC. This ELB is not managed by or even known to Crossplane, and is not deleted along with the EKS cluster. This means that the Helm Release (and therefore its Service) must be deleted before the EKS cluster in order to ensure the EKS cluster’s controllers trigger deletion of the ELB. If it is not, the ELB will be orphaned, costing money and blocking the deletion of the VPC it was created in.

These problems are exacerbated by the fact that Kubernetes uses “background cascading deletion” by default. Put otherwise, the successful deletion of a parent resource is not dependent on the successful deletion of its children. Instead the parent resource will be deleted successfully and the children (i.e. resources that declare the parent as an owner reference) will later be deleted by the Kubernetes garbage collector. This makes it difficult to diagnose or even notice resources that become orphaned when their parent XR is deleted.

### How could Crossplane help solve your problem?

We recently put together [a document](https://docs.google.com/document/d/1Yu3GxNMJOOMWMjD6gEw5QtT8Fz5qQSyo18ZZFKqcqKw/edit#) that explores some options to ensure that dependencies are not deleted before their dependents. The leading proposal is to introduce a `Usage` type that models a dependency between two resources; a generic version of the existing `ProviderConfigUsage` type.

Most finalizer-based in-use protection implementations follow the approach outlined [in this article](https://www.ibm.com/cloud/blog/preserve-kubernetes-api-objects-while-they-are-in-use). Assume there are two resource types; Dependency and Dependent.

1. The Dependency controller adds a finalizer when a new Dependency is created.
2. When a Dependency is deleted the Dependency controller scans for (i.e. lists) Dependents.
3. If no Dependents depend on the Dependency, the Dependency controller removes the finalizer.

This approach relies on the fact that a dependency can have only one type of dependent. In cases where a dependency can have arbitrary types of dependents it’s possible to introduce an intermediate dependency object. This is how Crossplane ensures a ProviderConfig cannot be deleted when its in use:

1. The ProviderConfig controller adds a finalizer when a new ProviderConfig is created.
2. Each managed resource that uses a ProviderConfig creates a ProviderConfigUsage. The ProviderConfigUsage contains a reference to the managed resource and to the ProviderConfig.
3. When a ProviderConfig is deleted the ProviderConfig controller lists all ProviderConfigUsages.
4. If no ProviderConfigUsages reference the ProviderConfig, its controller removes the finalizer.

This approach could be extended to all Crossplane types. A new “Usage” type could be added to core Crossplane. Whenever an MR depended on another MR (or indeed a ProviderConfig, or an XR depending on another XR, etc) a “Usage” relationship would be manually created (e.g. composed as a resource in a Composition)  to reflect this relationship.

The Usage type would be “backed” by two finalizers (or [liens](https://github.com/kubernetes/enhancements/pull/2840), or admission control webhooks):

1. Controllers for resources that may be a dependency would add a finalizer to block their deletion until no Usages indicated they were an active dependency.
2. The controller for Usages would add a finalizer to block their deletion while their dependent resource still existed.

### Related Issues

- [x] https://github.com/crossplane/crossplane/issues/2439
- [ ] https://github.com/crossplane/crossplane/issues/2072
- [ ] https://github.com/crossplane/crossplane/issues/1770
- [ ] https://github.com/crossplane/crossplane-runtime/issues/350

**Comment by negz** [comment]
@bobh66 has prototyped this approach using https://kyverno.io, per the below gists:

* https://gist.github.com/bobh66/ace375505eb46ea8e3f29d318378b0c3
* https://gist.github.com/bobh66/715263901d7257e3593b3d753702a78f

**Comment by natereid72** [comment]
>Crossplane is built on Kubernetes, which leans into eventual consistency. When several things need to happen in a particular order the “Kubernetes way” is to specify the desired end state and rely on the relevant controllers to ensure the system eventually arrives at that state, largely through the use of constant reconciliation (aka retries).

There is StatefulSets within Kubernetes that also aims to address an `ordered` requirement. So not really breaking with the "Kubernetes way" imo.

**Comment by yebolenko** [comment]
any update on this issue?

**Comment by bobh66** [comment]
I posted a description of the Kyverno-based PoC here - https://github.com/crossplane/crossplane/discussions/4072

I have since ported this to a set of Python webhooks based on a customized `kopf` framework, which is a lot more efficient, and I added logic to trigger a delete of the depended-on object after the last Dependency is deleted, so it doesn't have to wait for the garbage collection backoff timer to expire, since that could be up to 1000 seconds.

I'm curious what the API would look like for a Crossplane-native implementation.  The UID tracking with labels is functional but not very elegant.

**Comment by negz** [comment]
> I'm curious what the API would look like for a Crossplane-native implementation.

@bobh66 my hunch is that we can come up with a single API type that satisfies two kinds of dependencies, i.e.:

* Don't delete dependency Y until dependent X is deleted. (Deletion ordering)
* Derive dependent field A from dependency field B. (Generic cross-resource references)

I sketched out the reference approach at https://github.com/crossplane/crossplane/issues/1770#issuecomment-1226632239. If we merge the two together I'm thinking it could be something like this:

```yaml
# A Reference used only for deletion ordering.
kind: Reference
metadata:
  name: release-uses-cluster
spec:
  # A 'blocking' reference prevents all of the 'to' (Cluster) resources from being
  # deleted until the 'from' resource (the Release) is gone.
  deletion: Blocking
  from:
    apiVersion: helm.crossplane.io/v1beta1
    kind: Release
    name: my-prometheus-chart
  to:
  - apiVersion: eks.upbound.io/v1beta1
    kind: Cluster
    name: my-cluster
```

```yaml
# A reference used only to derive field values.
apiVersion: crossplane.io/v1alpha1
type: Reference
metadata:
  name: subnet-to-vpc
spec:
  # A 'non-blocking' reference is the default. The VPC can be deleted before
  # the Subnet.
  deletion: NonBlocking
  from:
    apiVersion: network.aws.crossplane.io/v1alpha2
    kind: Subnet
    name: my-subnet
    # Note the optional fieldPath is specified here and below.
    fieldPath: spec.forProvider.vpcId
  to:
  # An array to support the case where a reference is derived from multiple
  # fields in multiple objects? This would likely end up pretty close to
  # Composition patches, with transforms etc. We'd probably want to support
  # selectors here too, not only explicit references.
  - apiVersion: network.aws.crossplane.io/v1alpha2
    kind: VPC
    name: my-vpc
    fieldPath: metadata.annotations[crossplane.io/external-name]
```

It would also be possible to use this API to model a deletion-blocking cross-resource reference.

Some questions come to mind:

* What API group would this live in? Today nothing lives in `crossplane.io` - always subdomains like `pkg.crossplane.io`.
* Is `Reference` the right type? I like that `Usage` has precedent in `ProviderConfigUsage`. It also feels more appropriate for the deletion ordering case to me, but less appropriate for the deriving-a-field-value (i.e. reference) case.

**Comment by negz** [comment]
> The UID tracking with labels is functional but not very elegant.

FWIW I think in your POC you're automatically creating the equivalent of the `Reference` type I sketched out above. I wouldn't do that; I'd require folks to create them explicitly.

**Comment by bobh66** [comment]
I thought about that but it seems like a lot of extra overhead - I would probably write a Composition Function to generate the references based on labels/annotations in the resources.

**Comment by maximilianbraun** [comment]
We have one case where we wrote a provider for a system which was drastically reliant on orders.  Its not too sophisticated, but it does the job. The solution only works within one provider.

> What API group would this live in? Today nothing lives in crossplane.io - always subdomains like pkg.crossplane.io.
For simplicity reasons i tend to `pkg.crossplane.io`. 

> Is Reference the right type? I like that Usage has precedent in ProviderConfigUsage. It also feels more appropriate for the deletion ordering case to me, but less appropriate for the deriving-a-field-value (i.e. reference) case.
I tend to `Usage`. Exactly with this argument. The purposes is to track usage between resources and to prevent deletion in the wrong order. And it follows the pattern. 

How do you want to derive if the dependency is `NonBlocking` or `Blocking`? We've worked a lot with tags in the types.

**Comment by bobh66** [comment]
I'm not an expert on kubernetes API query syntax, but I assume it's both possible and reasonably performant (or can be made to be) to query for all `Reference` objects that have `deletion: Blocking` and a specific resource in the `to` array?  The webhook would need to run this query, or it's equivalent, for every DELETE API call on a resource that matches the criteria of the webhook.

Similarly, for the generic cross-reference case, how would the reconciler know when to go looking for References to process?  Assuming (again) that the resource being processed is in the `to` array(s) of one or more References, the reconciler would need to look for all Reference objects that include a `to` entry for the resource being reconciled and process the "patching" from the source resource accordingly.

I imagine that there would need to be an indexer running on the Reference resource to facilitate these queries?

**Comment by bobh66** [comment]
>    What API group would this live in? Today nothing lives in crossplane.io - always subdomains like pkg.crossplane.io.

Seems like `apiextensions.crossplane.io` would make sense, since it already has Compositions and EnvironmentConfigs?

> Is Reference the right type? I like that Usage has precedent in ProviderConfigUsage. It also feels more appropriate for the deletion ordering case to me, but less appropriate for the deriving-a-field-value (i.e. reference) case.

I wasn't sure about Reference at first, but I think it's fine - it can be considered as a reference to an entire resource in the deletion case, and a reference to a specific resource attribute in the deriving-a-field-value case.

**Comment by turkenh** [comment]
> @bobh66 my hunch is that we can come up with a single API type that satisfies two kinds of dependencies, i.e.:
> 
> * Don't delete dependency Y until dependent X is deleted. (Deletion ordering)
> * Derive dependent field A from dependency field B. (Generic cross-resource references)

@negz I have some concerns about solving the two use cases with a single API type which I detailed [here](https://github.com/crossplane/crossplane/issues/4100#issuecomment-1568187121). I believe we can continue the discussion from there and close that ticket if we agree that solving them separately makes more sense.

I want to focus on solving the deletion ordering with a dedicated `Usage` type here.

The solution I want to propose is quite similar to what you've described above, with slight differences. We will have a generic `Usage` type as follows:

```yaml
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: Usage
metadata:
  name: release-uses-cluster
spec: 
  of: 
    # This could still be an array, however, considering existing use cases,
    # I believe we would rarely need arrays here and multiple `Usage` objects 
    # could be created instead of having this as an array.
    apiVersion: eks.upbound.io/v1beta1
    kind: Cluster
    name: my-cluster
    # implement selectors
  by:
    apiVersion: helm.crossplane.io/v1beta1
    kind: Release
    name: my-prometheus-chart
    # implement selectors
```

The Usage controller would put an annotation on the used resource as `crossplane.io/external-delete-blocked-by: <comma-separated-uuids-of-blockers>`. The managed reconciler would reject the deletion of the external resource when there is this annotation on a resource with a deletion timestamp. 

This is mostly not to introduce a type dependency from Providers to Crossplane but also to avoid listing/querying usage objects during the deletion of all managed resources. The proposed annotation feels native considering the existing `crossplane.io/external-create-*` annotations but with two caveats:
- Comma-separated list of UUIDs feels weird. However, it should be ok considering:
  - This annotation is only for controllers, and we don't expect users to interact with
  - We have enough room to keep around [7000 UUIDs](https://github.com/crossplane/crossplane/assets/9900707/802e05d7-dcc1-4924-a97b-ae46d9c51a36) which is a more than a reasonable limit for the use case here. 
- One controller putting an annotation on a resource managed by another controller. 
  - I don't think this is an anti-pattern though, and cannot think of any problems.

I've POC'd the above approach and verified it behaves as expected:

- https://github.com/turkenh/crossplane-runtime/pull/1
- https://github.com/turkenh/crossplane/pull/1

For providers, there is no change needed other than consuming the runtime changes.

**Comment by bobh66** [comment]
> The Usage controller would put an annotation on the used resource as crossplane.io/external-delete-blocked-by: <comma-separated-uuids-of-blockers>. The managed reconciler would reject the deletion of the external resource when there is this annotation on a resource with a deletion timestamp.

I think this is an interesting solution because it appears (to me) to be fundamentally no different than adding a finalizer to the resource to block deletion until the usage controller removes the finalizer when the last dependency has been deleted. 

Or it could even be a finalizer-per-dependent UUID to replace the comma-separated list of UUIDs, which I agree feels weird.

Note that storing only UUIDs makes debugging and troubleshooting difficult as there is no way to find a resource given only it's UUID.

The POC only appears to be implemented for Managed Resources - Composite resources are not included?  Dependencies on Composites are needed in order to block the deletion of the entire composite tree.

I'm a little concerned about how this would work with nested composites and foreground cascading deletion, since Crossplane delegates deletion to Kubernetes via ownerReferences.  If the Usage is only referring to the Composite and not to any child Composites or specific MRs, there is nothing to prevent the deletion of the child Composite/MRs while leaving the parent Composite blocked.

**Comment by turkenh** [comment]
Good points @bobh66! 

> The POC only appears to be implemented for Managed Resources - Composite resources are not included? Dependencies on Composites are needed in order to block the deletion of the entire composite tree.

TBH, I was too focused on the MR case and forgot about the dependencies on XRs. I believe we won't be able to proceed with annotations in this case, however, we could go with finalizers instead by keeping the logic of the Usage controller the same, except it would add/remove finalizers per dependency. 

> I think this is an interesting solution because it appears (to me) to be fundamentally no different than adding a finalizer to the resource to block deletion until the usage controller removes the finalizer when the last dependency has been deleted.

Agreed. I believe, I leaned on annotation because of the similarity of existing annotations for external-create-*, but completely forgot about dependencies on XRs.

> I'm a little concerned about how this would work with nested composites and foreground cascading deletion, since Crossplane delegates deletion to Kubernetes via ownerReferences. If the Usage is only referring to the Composite and not to any child Composites or specific MRs, there is nothing to prevent the deletion of the child Composite/MRs while leaving the parent Composite blocked.

I believe the usage of the finalizers might solve this case as well. But I will investigate further, especially not sure how garbage collection behaves for a resource with finalizer and foreground cascading deletion.

**Comment by bobh66** [comment]
Thanks @turkenh!  The main reason I ended up using the vadlating admission webhook in my PoC was that it was necessary to block the DELETE API call on the XR so that none of the child resources would be marked for deletion.  If the DELETE API call is not blocked then all of the child resources get marked for deletion, and the reconcilers would need to know about the dependency on the higher level XR in order to block the deletion of the child XRs/MRs as long as the parent XR is blocked.  That would have required reconciling the parent's blocking annotation/finalizer to all of the children, which got ugly very quickly.

An alternative would be to have Crossplane take over the deletion handling by removing the ownerReferences, but I think that opens up more problems than it solves, and it also breaks things like the `lineage` tool and ArgoCD tracking of resources.

GC will not collect a resource as long as there is a finalizer on it, but I don't know if that helps in the XR case since GC will still follow the ownerReferences and trigger deletion of the edge MRs and works its way back up the tree (in the foreground case).
I was not able to find a solution that did not involve blocking the DELETE API call, but that doesn't mean there isn't one.

**Comment by turkenh** [comment]
Thanks a lot for the hints @bobh66!

Let me try to summarize to make sure we are on the same page. 

There are indeed three distinct types of resources that we want to block the deletion of:

1. A Managed Resource.
   - We control the deletion logic; we have the most power.
   - We can use an annotation or finalizer (or whatever we want) and handle the logic in managed reconciler for all managed resources.
2. A Composite Resource with the (default) **background** deletion policy.
   - We rely on the garbage collector for deletion. Composed resources got the deletion request only after the Composite disappeared from the API. 
   - So, we can still rely on a finalizer-based approach to block the deletion of the whole tree of owned/composed resources.
3. A Composite Resource but with the **foreground** deletion policy.
   - We still rely on the garbage collector for deletion; however, composed resources got the deletion request immediately after the Composite got the deletion request, no matter whether it has additional finalizers or not.
   - Finalizer-based approaches wouldn't work out of the box unless we implemented ugly workarounds like propagating special finalizers down to composed resources.

So, we have two possible solutions to cover all the scenarios:

A. Do not rely on the garbage collector by removing owner references to delete composed resources and implement our custom deletion logic.
  - This _"opens up more problems than it solves, and it also breaks things like the lineage tool and ArgoCD tracking of resources."_ as you said.

B. Block the delete request with the help of a webhook.

> I was not able to find a solution that did not involve blocking the DELETE API call, but that doesn't mean there isn't one.

So, currently, I am on the same page as you and cannot think of any obvious/easy solution other than blocking the deletion request.

That said, I've pivoted my POC to a webhook based solution and it turned out to be a pretty clean solution:
- It only requires changes in the core Crossplane (no changes required in the providers).
- Existing machinery for webhooks makes it very easy to add the new one. 
- Webhook logic is [pretty simple and straightforward](https://github.com/turkenh/crossplane/pull/2/files#diff-5f6ee04de3407304093dc0aafefebaf48280800efb4404ebec674e55ddc3e9d4R99-R101) with the help of indexer fields, even if the used field is an array.
- For now, I registered [all DELETE calls to all resources](https://github.com/turkenh/crossplane/pull/2/files#diff-fc1af1d389087d6aacc9396b054e415de39334000329135f0833c12a5eb11e3dR18-R25), but we can quickly implement registering only XP managed/composite/claim resources or register on demand when a `Usage` resource created.

Please see: https://github.com/turkenh/crossplane/pull/2

**Comment by turkenh** [comment]
> With Background propagation policy there is no ordering at all, or at least none that is easily observed. The GC generates the resource tree based on the ownerReferences and then proceeds to make DELETE API calls to all of the resources at the same time. It does not wait for any specific resource to be deleted before triggering the deletion of the "next" resource. I think of it as "pulling the rug out" - everything falls down at once, and you end up with a bunch of MRs (that take time to delete) and no XRs, which are effectively deleted immediately as the reconcilers have little to do for deletion.

But this process starts only after the parent object disappeared, doesn't it? This is how I experimented and came to this conclusion:

```
> kubectl create deployment nginx --image nginx --replicas=3

> kubectl edit deploy nginx
# and add a random finalizer manually, to block the resource disappearing from the API, e.g. `- finalizer.extensions/v1beta1`, not any random string accepted.

> kubectl delete deploy nginx
```

At this point, if I check replicasets or pods, notice that they don't get any deletion request yet. This happens only after I remove the finalizer that I manually added.

> One other thing you may find is that with large nested composites deletion can take a long time because of the exponential backoff timer that GC uses when the DELETE API call is rejected. The max backoff is 1000 seconds, meaning that once the "Usage" indicator of the object has been deleted. it could take up to 1000 seconds for GC to re-attempt deletion of the resource that was being blocked.
> 
> I solved this in my PoC by adding logic to re-trigger deletion of the resource when the dependency is deleted, and it speeds up the process significantly. The delete request may be rejected if other Usage references still exist, but when the last one is deleted it will cause the delete process to continue immediately instead of waiting for up to 15+ minutes.

Cool, I will keep this in mind 🙏

**Comment by turkenh** [comment]
We had a chance to sync with @negz and @sttts on this topic in person. Both had some concerns about building a webhook-based solution because of the following: 
- Clients should be able to declare their desire for deletion without getting an error and/or worrying about retrying it later. Our controller should be able to complete the deletion.
- Returning an error to (rejecting) a DELETE API call might cause unexpected behavior depending on the implementation of the client. The garbage collector backing off up to 1000 seconds is already a good example of this.

So, we did a quick brainstorming around a solution without webhook but covering all the cases, including XR with Foreground deletion. Currently, we are leaning toward a solution where dependencies to XRs are being resolved as dependencies to individual composed resources where new `Usage` resources will be created by the XR controller (or Usage controller) for composed resources. After ensuring that there is always a `Usage` object for managed resources either used directly or indirectly as part of an XR, the managed reconciliation will simply postpone the deletion of the external resource until no usage is left.

Let's take the [XCluster](https://github.com/upbound/platform-ref-aws/blob/1149618f72e1eb32b5119eab776a3c4941135dba/package/cluster/composition.yaml#L9) composition in platform-ref-aws which composes `XNetwork`, `XEKS` and `XServices` composites where `XServices` contains two release to be deployed into `XEKS`. So, there is a dependency from `XServices` to `XEKS` and we will define a `Usage` as follows:

```
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: Usage
spec: 
  of: 
  - apiVersion: aws.platformref.upbound.io/v1alpha1
    kind: XEKS
    nameSelector:
      matchControllerRef: true
  by:
    apiVersion: aws.platformref.upbound.io/v1alpha1
    kind: XServices
    nameSelector:
      matchControllerRef: true
```

The XR controller will list usages on the composite its reconciling, and if exist, it will resolve it to the resources it composes and create another Usage type(s) per using resource. In this specific example, during the reconciliation of `XEKS`, the composite controller will create the following `Usage` in addition to the already existing one:

```
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: Usage
spec: 
  of: 
  - apiVersion: iam.aws.upbound.io/v1beta1
    kind: Role
    name: <name-of-the-composed>
  - apiVersion: iam.aws.upbound.io/v1beta1
    kind: RolePolicyAttachment
    name: <name-of-the-composed>
  - apiVersion: eks.aws.upbound.io/v1beta1
    kind: Cluster
    name: <name-of-the-composed>
  - apiVersion: eks.aws.upbound.io/v1beta1
    kind: ClusterAuth
    name: <name-of-the-composed>
  - ...
  by:
    apiVersion: aws.platformref.upbound.io/v1alpha1
    kind: XServices
    name: service-...
```

If there were nested XRs, they would be resolved as well, resulting in further `Usage`s. 

Now, when the `XEKS` gets a deletion request, composed resources will get the same, but the existence of `Usage` resource will prevent them from being deleted. 

There are still some details that need to be ironed out and some questions like which controller should make the resolution. I believe I will have a better understanding after doing some POC.

**Comment by bobh66** [comment]
@turkenh this is an interesting solution, it is essentially reconciling the `Usage` resource down the tree to all of the "child" resources of the one that has the dependency declared on it.  I looked at doing something similar with finalizers until that effort was shelved in favor of the proposal for in-use protection using `Liens` - https://github.com/kubernetes/enhancements/pull/2840

Whatever the implementation for `Liens` ends up being it seems like the desired end result would be the same - reject the DELETE API call until the `Lien` no longer exists.

A couple of concerns:

- The reconciliation of the `Usage` resource will have to be continuous - there may be many levels of nested composites under the resource that is depended-on and not all of the resources will necessarily exist when the initial `Usage` resource is created and reconciled.  As new resources under the dependency are created they will need to have `Usage` resources added to them.

- Multiple `Usage` dependencies can exist on the same "parent" resource which will result in multiple sets of `Usage`s created for all of the children - the number of `Usage` objects will expand rapidly into the hundreds or more for a single deployment (in our use case).

I'm not opposed to the solution, just concerned about the increased complexity and what it means for performance and debugging.

Thanks for all of your work on this!

**Comment by bobh66** [comment]
> Clients should be able to declare their desire for deletion without getting an error and/or worrying about retrying it later. Our controller should be able to complete the deletion.

Thinking about it some more, is this something that Crossplane really wants to take on?  Why would the DELETE handling for APIs that Crossplane defines behave differently to external clients than other Kubernetes APIs?

The cloud providers will reject the delete of a resource that has other resources depending on it - for example a VPC that has subnets defined, and it's up to the client to try again.  They don't treat it as "delete this thing whenever it is ok to do so".  That could result in pending deletes hanging around which I don't think would be desirable.

I think it's reasonable to expect that the "client" will mostly be dealing with the Claim or a top-level Composite, and so would be a lot less likely to experience a rejected DELETE API call than the garbage collector.  Assuming that there are no _externally defined_ `Usage` resources referring to the Claim/Composte, the user will never see a rejected delete call.

The user/client should only experience rejected deletes when there is an externally defined `Usage` on a Claim or top-level Composite.  In this case the rejection of the DELETE API call reminds the user that an _external_ dependency exists and should be removed first.  I definitely _don't_ want Crossplane to mark everything in that claim as deleted and wait for that _external_ dependency to be deleted by a second API call which may never come.

> Returning an error to (rejecting) a DELETE API call might cause unexpected behavior depending on the implementation of the client. The garbage collector backing off up to 1000 seconds is already a good example of this.

Isn't that the client's responsibility to define and document the behavior?  And the user's responsibility to understand the behavior of the client?  I agree that the 1000 second backoff in the GC can be annoying but from their perspective I don't know what else they could do.

I think the bottom line is that the DELETE call _should_ be rejected in the case of a Claim/Composite having a `Usage` reference because it implies an _external_ dependency which is outside of the specific composite.  That composite should not sit around in a pending-deletion state waiting for that external dependency to be removed.

I'd be happy to discuss further if needed.

**Comment by negz** [comment]
> I'm not opposed to the solution, just concerned about the increased complexity and what it means for performance and debugging.

FWIW this is my feeling too. Resolving usage of an XR to usages of its composed resources feels like the _least bad_ option we've identified so far but I don't feel great about it.

> Thinking about it some more, is this something that Crossplane really wants to take on? Why would the DELETE handling for APIs that Crossplane defines behave differently to external clients than other Kubernetes APIs?

We're trying to match the behaviour of deletion in Kubernetes, which AFAICT currently is typically "declarative" and eventually consistent.

I'm aware of two pieces of prior art - Kubernetes [PVC in-use protection](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#storage-object-in-use-protection) and Crossplane `ProviderConfig` in-use protection (i.e. `ProviderConfigUsage`). In both of these cases deletion is declarative and eventually consistent; the API server accepts the desired state (i.e. to delete the resource in question) and does that when it can - i.e. when the resource in question is no longer being used by any other resource.

Interestingly, [the Liens proposal](https://github.com/kubernetes/enhancements/blob/8ebd03ba24517f375794d53ad1fbe75a211550b8/keps/sig-api-machinery/2839-in-use-protection/README.md) (which is not yet accepted) proposes that the API server respond with an HTTP 409 when Liens are present. So it does seem ambiguous whether "declarative deletes" are actually desired upstream or just a limitation of the machinery available to implement them (i.e. finalizers).

I agree this is different from how the cloud providers handle deletion. It does feel pretty in-line with how Kubernetes and Crossplane work in other ways. Consider for example _creating_ an MR that is dependent on another MR in order for creation to work at the cloud-provider level. We don't reject the MR create call if it's not possible; instead we accept it and keep trying to create the external resource until we can. It does seem to me that it would be ideal for deletes to work this way too. Making deletes declarative and eventually consistent feels like it makes life easier for clients by letting them avoid having to constantly retry operations.

**Comment by negz** [comment]
I dropped a comment on the Liens proposal capturing my concerns WRT returning a 409: https://github.com/kubernetes/enhancements/pull/2840/files#r1224844740

**Comment by bobh66** [comment]
> Considering that there is no clear agreement in the upstream

I think upstream has left it to the user to decide - if the client sets `propagationPolicy: background` in the DELETE API call (the default) the server will handle it as a declarative delete and return immediately after marking the specified object for deletion.  If the user sets `propagationPolicy: foreground` then the server will delete all of the child objects before deleting the specified object, and then return a response.  That is a more imperative workflow that can be blocked/delayed.  I think it's good to have options and not be bound to a single implementation that may not be valid for all use cases.

**Comment by maximilianbraun** [comment]
> Considering that there is no clear agreement in the upstream, I think it could be good to choose a practical approach and start with the simpler solution, namely the webhook based solution. I believe it's better to start with the simple option, and then, if we find out that it's not enough, we can think about changing to a more complex solution. I won't expect an API change even if we change the underlying implementation later, except getting rejected deletes vs. not.

How well would that work with GitOps tools like Argo CD and FluxCD? I imagine them pretty fast running into a situation where you need to manually retrigger syncs because their mechanism back-off. (That was my experience in different situations)

**Comment by turkenh** [comment]
> I think upstream has left it to the user to decide - if the client sets propagationPolicy: background in the DELETE API call (the default) the server will handle it as a declarative delete and return immediately after marking the specified object for deletion. If the user sets propagationPolicy: foreground then the server will delete all of the child objects before deleting the specified object, and then return a response. That is a more imperative workflow that can be blocked/delayed. I think it's good to have options and not be bound to a single implementation that may not be valid for all use cases.

Not sure if this is directly related to the problem here, I mean "how to handle deletions when there are dependencies? e.g. reject delete call vs accept delete call but delay until no dependency left". As you already know, it is about configuring whether the parent resource should live in the API or not until all child resources are deleted.
Also, I assume this _wait until resource deleted before sending the response_ logic is something implemented on the client side (e.g. [kubectl](https://github.com/kubernetes/kubernetes/issues/42594)) rather than the server side.

However, I do agree that we should look for ways to leave this to users as an option in general but would rather prefer to invest in that once we see there is a clear demand as it would be more complex than having either one of the options.

**Comment by bobh66** [comment]
> How well would that work with GitOps tools like Argo CD and FluxCD? I imagine them pretty fast running into a situation where you need to manually retrigger syncs because their mechanism back-off. (That was my experience in different situations)

I would expect that Argo and Flux are only dealing with Claims.  As long as there are no external dependencies pointing at the Claims Argo/Flux will never get a rejected DELETE call.

We are using Argo with the PoC implementation of the webhook and have not seen any problems because all of the dependencies that trigger rejects are inside the composites.  The client never sees those interactions.

If someone needs external dependencies between claims those will need to be bundled into nested composites or handled at a higher level of orchestration.

**Comment by bobh66** [comment]
> Also, I assume this wait until resource deleted before sending the response logic is something implemented on the client side (e.g. https://github.com/kubernetes/kubernetes/issues/42594) rather than the server side.

Yes - after more testing I'm not sure why I was seeing the delay on the API response earlier - it may have been related to the webhook processing.

I just wanted to point out that upstream is not forcing a specific delete behavior, they provide options to allow for different use cases and requirements.  I fully agree that starting simple with the webhook solution is the best path forward, and additional use cases and requirements can be addressed as they are defined.

**Comment by turkenh** [comment]
I am throwing out two more potential alternatives, not necessarily because they are my preferred choices, but in the hope of stimulating further ideas 🙂 

**Option C:** Have a mutating webhook instead of a validating one, and convert all delete calls to all XP resources to a deferred deletion state. Clients will see the response to delete call as a success and do not worry about retrying. Then Crossplane's MR and XR controllers will handle this interim state and trigger the deletion only after no Useage's left. The caveat is we would need to store more information for this interim state, like propagationPolicy or gracePeriodSeconds. Another potential problem is the diff between users desired state (i.e. the resource has a deletionTimestamp) and the actual state despite a successful delete request.

**Option D:** Leverage owner references to prevent deletion of the composed resources when the composite is deleted with deletionPolicy foreground. The composite controller will add the "`Usage` object(s) using the composite" as other owner(s) of the composed resources. This way, the garbage collector won't touch those resources, and we won't have to create additional Usage's by resolving all the way down to the leafs. Doing this only for direct dependencies would be sufficient, and we won't need to continue propagating down for nested composites.
